components/tabs/src/storage.rs

```diff
@@ -188,27 +211,80 @@ impl TabsStorage {
+                        // could happen - in most cases though, it will be due to a disconnected client -
+                        // so we really should consider just dropping it? (Sadly though, it does seem
+                        // possible it's actually a very recently connected client, so we keep it)
+                        log::info!(
+                        "Storing tabs from a client that doesn't appear in the devices list: {}",
+                        id,
+                    );
+                        ClientRemoteTabs::from_record(id, last_modified, record)
+                    };
+                    crts.push(crt);
                 }
+                Some(crts)
+            }
+        }
+    }
+
+    pub fn remove_stale_clients(&mut self) -> Result<()> {
+        let last_sync = self.get_meta::<i64>(schema::LAST_SYNC_META_KEY)?;
+        if let Some(conn) = self.open_if_exists()? {
+            if let Some(last_sync) = last_sync {
+                let tx = conn.unchecked_transaction()?;
+                // Get rid of anything older than 3 weeks of our last sync
+                tx.execute_cached(
+                    "DELETE FROM tabs WHERE last_modified <= :last_sync - :ttl",
+                    rusqlite::named_params! {
+                        ":last_sync": last_sync,
+                        ":ttl": TTL_3_WEEKS,
+                    },
+                )?;
+                tx.commit()?;
             }
         }
+        Ok(())
     }
 }
 
 impl TabsStorage {
     pub(crate) fn replace_remote_tabs(
         &mut self,
-        new_remote_tabs: Vec<ClientRemoteTabs>,
+        // This is a tuple because we need to know what the server reports
+        // as the last time a record was modified
+        new_remote_tabs: Vec<(TabsRecord, ServerTimestamp)>,
     ) -> Result<()> {
         let connection = self.open_or_create()?;
         let tx = connection.unchecked_transaction()?;
-        // delete the world - we rebuild it from scratch every sync.
-        tx.execute_batch("DELETE FROM tabs")?;
 
-        for crt in new_remote_tabs {
+        // For tabs it's fine if we override the existing tabs for a remote
+        // there can only ever be one record for each client
+        for remote_tab in new_remote_tabs {
+            let record = remote_tab.0;
+            let last_modified = remote_tab.1;
             tx.execute_cached(
-                "INSERT INTO tabs (payload) VALUES (:payload);",
+                "INSERT OR REPLACE INTO tabs (guid, record, last_modified) VALUES (:guid, :record, :last_modified);",
                 rusqlite::named_params! {
-                    ":payload": serde_json::to_string(&crt).expect("tabs don't fail to serialize"),
+                    ":guid": &record.id,
+                    ":record": serde_json::to_string(&record).expect("tabs don't fail to serialize"),
+                    ":last_modified": last_modified.as_millis()
```

Comment:
So each client has exactly one tab record (which contains all the tabs) which is stored as a json (in `record`, though maybe I should change the name to avoid this confusion). So if a client has an updated tab list -- it would only ever overwrite it's own row. Thus preventing any dangling records.

Also I have a `remove_stale_clients()` that removes any stale clients > 3 weeks during sync (in `apply_incoming`) if a client Id changes and/or never connects again to prevent the DB to be filled with any stragglers

Reply:
Ahh, I should have said delete the client records that are no longer present.  Looks like that's covered too though.